runtime/src/accounts_db.rs

```diff
@@ -883,46 +876,74 @@ impl AccountsDB {
+        let mut total_alive_count = 0;
+        let mut all_stored_accounts: Vec<(
+            Slot,
+            Pubkey,
+            Account,
+            usize,
+            (AppendVecId, usize),
+            u64,
+        )> = vec![];
+        let mut storages = self.shrink_candidates.lock().unwrap();
+        while let Some(storage) = storages.pop() {
+            trace!("shrink_stale_storage: storage: {:?}", storage);
+
+            let storage = storage.upgrade();
+            if storage.is_none() {
+                continue;
+            }
+            let storage = storage.unwrap();
+
+            let alive_count = storage.count();
+            let mut stored_accounts = vec![];
+            let mut start = 0;
+            // XX change to get_account(start, None)
+            while let Some((account, next)) = storage.accounts.get_account(start) {
+                stored_accounts.push((
+                    account.meta.slot,
+                    account.meta.pubkey,
+                    account.clone_account(),
+                    next - start,
+                    (storage.id, account.offset),
+                    account.meta.write_version,
+                ));
+                start = next;
+            }
 
-        let mut stored_accounts = vec![];
-        {
-            let storage = self.storage.read().unwrap();
-            if let Some(stores) = storage.0.get(&slot) {
-                let mut alive_count = 0;
-                for store in stores.values() {
-                    alive_count += store.count();
-                    let mut start = 0;
-                    while let Some((account, next)) = store.accounts.get_account(start) {
-                        stored_accounts.push((
-                            account.meta.pubkey,
-                            account.clone_account(),
-                            next - start,
-                            (store.id, account.offset),
-                            account.meta.write_version,
-                        ));
-                        start = next;
-                    }
-                }
-                if (alive_count as f32 / stored_accounts.len() as f32) >= 0.80 {
-                    trace!(
-                        "shrink_stale_slot: not enough space to shrink: {} / {}",
-                        alive_count,
-                        stored_accounts.len()
-                    );
-                    return;
-                }
+            if stored_accounts.len() > 1000
+                && (alive_count as f32 / stored_accounts.len() as f32) >= 0.80
+            {
+                trace!(
+                    "shrink_stale_storage: not enough space to shrink: {} / {}",
+                    alive_count,
+                    stored_accounts.len()
+                );
+                continue;
+            }
+            total_alive_count += alive_count;
+            all_stored_accounts.extend(stored_accounts);
+
+            if total_alive_count > 100 {
```

Comment:
This is tooo rough. But I intend a validator to hold many appendvecs of minimum size of 2MiB or 4MiB?